src/Compilers/CSharp/Portable/Symbols/DelegateCaches/FullyConcreteChecker.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    /// <summary>
+    /// Checks whether a type or a method is fully constructed with fully concrete types as opposed to yet another generic parameter.
+    /// </summary>
+    internal class FullyConcreteChecker : CSharpSymbolVisitor<bool>
+    {
+        public static FullyConcreteChecker Instance { get; } = new FullyConcreteChecker();
+
+        public override bool VisitTypeParameter(TypeParameterSymbol symbol) => false;
+
+        public override bool VisitAlias(AliasSymbol symbol) => true;
+
+        public override bool VisitArrayType(ArrayTypeSymbol symbol) => Visit(symbol.ElementType);
+
+        public override bool VisitNamedType(NamedTypeSymbol symbol)
+        {
+            if (symbol.IsAnonymousType)
+            {
+                foreach (var anonPropType in AnonymousTypeManager.GetAnonymousTypePropertyTypes(symbol))
+                {
+                    if (!Visit(anonPropType))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            foreach (var typeArg in symbol.TypeArguments)
```

Comment:
Custom modifiers of type arguments can probably refer to type parameters. We should visit them too. And in the other visitor as well.